crates/sqlexec/src/background_jobs.rs

```diff
@@ -0,0 +1,326 @@
+                None => return Ok(()),
+            }
+        };
+        let _ = handle.await;
+
+        Ok(())
+    }
+
+    /// Add a new background job.
+    pub fn add(&self, name: impl ToString, options: BackgroundJobOptions) -> Result<()> {
+        let job = BackgroundJob::new(name, options)?;
+        self.send
+            .send(ChannelMessage::NewJob(job))
+            .map_err(|e| ExecError::ChannelSendError(Box::new(e)))?;
+        Ok(())
+    }
+}
+
+/// A struct representing a batched background job. These jobs are represented
+/// by their "name" and batched until decided that they need to be started.
+struct BackgroundJob {
+    name: String,
+    inner: Box<dyn BgJob>,
+}
+
+impl BackgroundJob {
+    /// Creates a new background job. The job will be batched or not based on
+    /// its properties.
+    pub fn new(name: impl ToString, options: BackgroundJobOptions) -> Result<Self> {
+        let inner = match options {
+            BackgroundJobOptions::Debug(debug_opts) => {
+                Box::new(BackgroundJobDebug::new(debug_opts))
+            }
+        };
+
+        Ok(Self {
+            name: name.to_string(),
+            inner,
+        })
+    }
+}
+
+#[async_trait]
+pub trait BgJob: Send + Sync {
+    /// The job should start at this instant (and not before this).
+    fn start_at(&self) -> Instant;
+
+    /// Main function that runs the job.
+    async fn start(&self) -> Result<()>;
+}
+
+pub enum BackgroundJobOptions {
+    Debug(BackgroundJobOptionsDebug),
+}
+
+pub struct BackgroundJobOptionsDebug {
+    pub batched: bool,
+    pub counter: Arc<AtomicUsize>,
+}
+
+struct BackgroundJobDebug {
+    start_at: Instant,
+    counter: Arc<AtomicUsize>,
+}
+
+impl BackgroundJobDebug {
+    fn new(opts: BackgroundJobOptionsDebug) -> Self {
+        let start_at = if opts.batched {
+            Instant::now() + Duration::from_secs(1)
+        } else {
+            Instant::now()
+        };
+        Self {
+            start_at,
+            counter: opts.counter,
+        }
+    }
+}
+
+#[async_trait]
+impl BgJob for BackgroundJobDebug {
+    fn start_at(&self) -> Instant {
+        self.start_at
+    }
+
+    async fn start(&self) -> Result<()> {
+        self.counter.fetch_add(1, Ordering::Relaxed);
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
+    async fn test_background_jobs() {
+        let jobs = JobRunner::new();
+
+        let counter = Arc::new(AtomicUsize::new(0));
+
+        // Start an instantaneous job.
+        jobs.add(
+            "test_instantaneous",
+            BackgroundJobOptions::Debug(BackgroundJobOptionsDebug {
+                batched: false,
+                counter: counter.clone(),
+            }),
+        )
+        .unwrap();
+
+        // TODO: Add mechanism to wait for a certain job to finish. Sleeps lead
+        // to flaky tests.
+        tokio::time::sleep(Duration::from_millis(10)).await;
```

Comment:
We could use a channel instead of an atomic, then await the recv side.

Reply:
Yeah, just need to add another set of channels here (for reverse communication).